modules/metrics/src/main/java/org/apache/ignite/internal/metrics/AbstractMetricSource.java

```diff
@@ -71,32 +64,34 @@ protected AbstractMetricSource(String name) {
      *
      * @return Metrics holder instance if metrics are enabled, otherwise - {@code null}.
      */
-    public final T holder() {
+    public final @Nullable T holder() {
         return holder;
     }
 
     /**
-     * Method is responsible for creation of appropriate holder instance in underlying implementations.
+     * Method is responsible for creation of appropriate <b>immutable</b> holder instance in underlying implementations.
      *
-     * @return New instance of metrics holder that must implements {@link Holder} interface.
+     * @return New <b>immutable</b> instance of metrics holder that must implements {@link Holder} interface.
      */
     protected abstract T createHolder();
 
-    /** {@inheritDoc} */
-    @Override public final MetricSet enable() {
-        MetricSetBuilder bldr = new MetricSetBuilder(name);
+    @Override
+    public final @Nullable MetricSet enable() {
+        T newHolder = createHolder();
 
-        T hldr = createHolder();
+        if (HOLDER_FIELD_UPD.compareAndSet(this, null, newHolder)) {
+            var metricSetBuilder = new MetricSetBuilder(name);
 
-        init(bldr, hldr);
+            init(metricSetBuilder, newHolder);
```

Comment:
Now the return of uninitialized holder by `holder()` method became possible (see `init()`). I would leave it as it was before.

Reply:
I tried to fix it as discussed in private.